src/audio/copier/copier.c

```diff
@@ -54,10 +54,45 @@ DECLARE_SOF_RT_UUID("copier", copier_comp_uuid, 0x9ba00c83, 0xca12, 0x4a83,
 
 DECLARE_TR_CTX(copier_comp_tr, SOF_UUID(copier_comp_uuid), LOG_LEVEL_INFO);
 
-static pcm_converter_func get_converter_func(const struct ipc4_audio_format *in_fmt,
-					     const struct ipc4_audio_format *out_fmt,
-					     enum ipc4_gateway_type type,
-					     enum ipc4_direction_type);
+/* In sof normal format conversion path, sample size should be equal
+ * to container size except format of S24_LE. In ipc4 case, sample
+ * size can be different with container size. This function is used to
+ * check conversion mode.
+ */
+static bool use_no_container_convert_function(enum sof_ipc_frame in,
+					      enum sof_ipc_frame valid_in_bits,
+					      enum sof_ipc_frame out,
+					      enum sof_ipc_frame valid_out_bits)
+{
+	/* valid sample size is equal to container size, go normal path */
+	if (in == valid_in_bits && out == valid_out_bits) {
+		if (in == SOF_IPC_FRAME_S24_3LE || out == SOF_IPC_FRAME_S24_3LE)
+			return false;
+
+		return true;
+	}
+
+	return false;
+}
+
+pcm_converter_func get_converter_func(const struct ipc4_audio_format *in_fmt,
+				      const struct ipc4_audio_format *out_fmt,
+				      enum ipc4_gateway_type type,
+				      enum ipc4_direction_type dir)
+{
+	enum sof_ipc_frame in, in_valid, out, out_valid;
+
+	audio_stream_fmt_conversion(in_fmt->depth, in_fmt->valid_bit_depth, &in, &in_valid,
+				    in_fmt->s_type);
+	audio_stream_fmt_conversion(out_fmt->depth, out_fmt->valid_bit_depth, &out, &out_valid,
+				    out_fmt->s_type);
+
+	/* check container & sample size */
+	if (use_no_container_convert_function(in, in_valid, out, out_valid))
+		return pcm_get_conversion_function(in, out);
+	else
+		return pcm_get_conversion_vc_function(in, in_valid, out, out_valid, type, dir);
+}
```

Comment:
why are you moving this part of the code? It makes it harder to review the actual change which is simply to expose those 3 functions

Reply:
which code convention?

Reply:
haha, normally, sub-function should be in upper part, and it will be called in lower in the source file.
I understand this is not must, since header file already declared this function prototype, and it can be used earlier before the real code, @marc-hb

Reply:
> normally, sub-function should be in upper part, and it will be called in lower in the source file.

Even with `static` functions you can use a forward declaration. I'm not aware of any code convention about this. It certainly does not deserve this git noise:

> why are you moving this part of the code? It makes it harder to review the actual change which is simply to expose those 3 functions